feat(shared-data): add deck schema v6

sfoster1 opened the pull request:
This schema contains a concept of "locating features" for addressable areas. These locating features are things that constrain the position of labware loaded on top of the addressable area. The presence of a locating feature indicates that the addressable area can contain deck items. Which locating feature is available constrains which labware can be loaded; different labware define different locatable features (for instance, their back left or front left corners, or the outer surface of one of their wells).

Locating feature names are not constrained to enums because we expect a fairly wide number of them.

## To come out of draft
- [ ] positions for modules that are not currently in

## risk
low, because nothing should be using this (purposeful, will be followup prs)

Closes EXEC-80

sfoster1 committed:
feat(shared-data): add deck schema v6

This schema contains a concept of "locating features" for addressable
areas. These locating features are things that constrain the position of
labware loaded on top of the addressable area. The presence of a
locating feature indicates that the addressable area can contain deck
items. Which locating feature is available constrains which labware can
be loaded; different labware define different locatable features (for
instance, their back left or front left corners, or the outer surface of
one of their wells).

Locating feature names are not constrained to enums because we expect a
fairly wide number of them.

Closes EXEC-80

sfoster1 committed:
Merge branch 'edge' into exec-80-deck-slot-defs-for-locating-features

sfoster1 committed:
don't need unitvector anymore

sfoster1 committed:
fixup: keep the version default

sfoster1 committed:
fixup: add missing stacker configuration

SyntaxColoring commented on shared-data/deck/schemas/6.json:
```diff
@@ -0,0 +1,352 @@
+        "zDimension": { "$ref": "#/definitions/positiveNumber" }
+      }
+    },
+    "locatingFeature": {
+      "type": "object",
+      "required": ["locatingFeatureId", "offsetVector"],
+      "properties": {
+        "locatingFeatureId": {
+          "type": "string",
+          "description": "The name of a kind of locating feature that will be looked up in labware. Common ones are backLeft, frontLeft, wellA1, center, but others may be used."
+        },
+        "offsetVector": {
+          "$ref": "#/definitions/xyzArray",
+          "description": "A vector from the origin of the thing that has the locating feature (i.e. an addressable area) to the key point of the locating feature."
+        }
+      }
+    }
```
> Locating feature names are not constrained to enums because we expect a fairly wide number of them.

I agree with not constraining them to enums, for versioning churn reasons, but I do think we'll need some central documented source that lists them all out and describes their semantics.

Like, what if we did something roughly like this:

```json
"anyOf": [
  { "const": "backLeft": "description": "..."},
  { "const": "center": "description": "..."},
  { "type": "string", "description": "reserved for future expansion" }
]
```

Alternatively, I guess this documentation could live in the labware definition schema. I don't have a preference.

sfoster1 replied:
Oh definitely! I was trying to remember how to do this and couldn't haha. I'll add that.

SyntaxColoring commented on shared-data/js/__tests__/deckSchemas.test.ts:
```diff
@@ -6,25 +6,27 @@ import glob from 'glob'
 const fixtureGlob = path.join(__dirname, '../../deck/fixtures/3/*.json')
 const defGlob = path.join(__dirname, '../../deck/definitions/3/*.json')
 const defV4Glob = path.join(__dirname, '../../deck/definitions/4/*.json')
+const defV5Glob = path.join(__dirname, '../../deck/definitions/5/*.json*')
+const defV6Glob = path.join(__dirname, '../../deck/definitions/6/*.json*')
```
What are the trailing `*`s for?

sfoster1 replied:
that's definitely a typo that got copy pasted

SyntaxColoring commented on shared-data/deck/definitions/6/ot2_standard.json:
```diff
@@ -0,0 +1,508 @@
+        "areaType": "slot",
+        "offsetFromCutoutFixture": [0.0, 0.0, 0.0],
+        "locatingFeatures": [
+          {
+            "locatingFeatureId": "frontLeft",
+            "offsetVector": [0.0, 86.0, 0.0]
+          },
+          {
+            "locatingFeatureId": "backLeft",
+            "offsetVector": [0.0, 0.0, 0.0]
+          }
```
Either:

* The slot origin is the front left, which means `frontLeft` should have an offset of `[0, 0, 0]` and `backLeft` should have an offset of `[0, 86, 0]`
* Or the slot origin is the back left, which means `frontLeft` should have an offset of `[0, -86, 0]` and `backLeft` should have an offset of `[0, 0, 0]`

Right?

sfoster1 replied:
...yes. I will fix this.

sfoster1 replied:
The thing I'm going to do is keep the slot origin as the front left. This is fine because actually what we're saying is that the AA origin is the point indicated by the `offsetFromCutoutFixture` vector, and all coordinates are relative to that. Conveniently, keeping that vector the same means that we can avoid changing some app code.

SyntaxColoring reviewed (commented):
Looks great so far. Very excited for this.

sfoster1 committed:
update: make defs a little more complex

sfoster1 committed:
remove incorrect z=0 notes

sfoster1 committed:
ah, i see

SyntaxColoring committed:
Delete accidental trailing wildcards.

SyntaxColoring committed:
Fix DeckSchemaId.

Remove the duplicate entry for v4 and add the missing entries for v5 and v6.

This appears unused, but we may as well correct it.

SyntaxColoring committed:
Rename AddressableArea -> AddressableAreaV4

This has a matingSurfaceUnitVector field, which does not exist in schema 6. Rename the type so it doesn't look more generally usable than it actually is.

SyntaxColoring reviewed (commented):
Looks good so far, other than whatever modules are missing.
